types/datum.go

```diff
@@ -917,12 +917,15 @@ func (d *Datum) convertToUint(sc *stmtctx.StatementContext, target *FieldType) (
 		val, err = ConvertFloatToUint(sc, d.GetMysqlSet().ToNumber(), upperBound, tp)
 	case KindBinaryLiteral, KindMysqlBit:
 		val, err = d.GetBinaryLiteral().ToInt(sc)
+		if err == nil && val > upperBound {
+			err = overflow(val, tp)
```

Comment:
Seems we should also consider the sql mode.
```
MySQL [test]> create table t(c tinyint);
Query OK, 0 rows affected (0.012 sec)

MySQL [test]> set @@sql_mode = "";
Query OK, 0 rows affected (0.001 sec)

MySQL [test]> insert into t values (0xFFFFFFFFFFF);
Query OK, 1 row affected, 1 warning (0.005 sec)

MySQL [test]> show warnings;
+---------+------+-------------------------------------------+
| Level   | Code | Message                                   |
+---------+------+-------------------------------------------+
| Warning | 1690 | constant 17592186044415 overflows tinyint |
+---------+------+-------------------------------------------+
1 row in set (0.001 sec)

MySQL [test]> select * from t;
+----------------+
| c              |
+----------------+
| 17592186044415 |
+----------------+
1 row in set (0.001 sec)
```

Reply:
`err = sc.HandleTruncate(err)` will covert Truncate As Warning, can we move it outside `func CastValue`?

Reply:
Maybe you should call `sc.HandleOverflow()`?

Reply:
The value of `t.c` in this PR is also `127`, which is the same as MySQL.
```
mysql> create table t(c tinyint);
Query OK, 0 rows affected (0.00 sec)

mysql> set @@sql_mode = "";
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t values (0xFFFFFFFFFFF);
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> select * from t;
+------+
| c    |
+------+
|  127 |
+------+
1 row in set (0.01 sec)
```
Why is your result `17592186044415` and is there anything wrong?
Please confirm it when it's convenient for you @lamxTyler .

Reply:
127 is correct.
the warning message is not compatible with MySQL.
it should be `Warning | 1264 | Out of range value for column 'a' at row 1`,
but we print `Warning | 1690 | constant 17592186044415 overflows tinyint`.
I'm tring to fix it now.